universum/modules/structure_handler.py

```diff
@@ -10,16 +11,23 @@
 ]
 
 
-def needs_structure(klass):
-    klass.structure_factory = Dependency(StructureHandler)
-    original_init = klass.__init__
+class HasStructure():
+    structure_factory: Dependency['StructureHandler']
+    def __init__(self):
+        structure: 'StructureHandler'
+
+
+def needs_structure(cls: Type) -> Type['HasStructure']:
```

Comment:
this loses information that resulting type is same class as incoming type, but preserves information about extra added fields. I'm not sure what would be the best way to preserve both.
The refactoring we discussed yesterday, where needs_structure becomes an interface/class resolves this issue.